src/main.c

```diff
@@ -1984,6 +1841,56 @@ void parse_blocks(int p, int q, int k, char* blocks, int size, State *s) {
     }
 }
 
+void place_block(int bp, int bq, int bx, int by, int bz, int bw) {
+    State *s = &g->players->state;
+    int bk = chunked_int(by);
+    Chunk *chunk = find_chunk(bp, bq, bk, 1);
+    if(chunk) {
+        parse_block(chunk,
+                    bx, by,
+                    bz, bw, s);
+        chunk->dirty = 1;
+        /* Make any chunk that owns an adjacent block dirty as well */
+        int r = 1;
+        for (int dx = -r; dx <= r; dx++) {
+            for (int dz = -r; dz <= r; dz++) {
+                for (int dy = -r; dy <= r; dy++) {
+                    int x = bx + dx;
+                    int y = by + dy;
+                    int z = bz + dz;
+                    int np = chunked_int(x);
+                    int nq = chunked_int(z);
+                    int nk = chunked_int(y);
+                    Chunk *c = find_chunk(np, nq, nk, 0);
+                    if (c) {
+                        c->dirty = 1;
+                    }
+                }
+            }
+        }
+        /* If the chunk is close to the player, force a chunk update */
+        if(chunk_near_player(bp, bq, bk, s, 10)) {
+            for (int dp = -r; dp <= r; dp++) {
+                for (int dq = -r; dq <= r; dq++) {
+                    for (int dk = -r; dk <= r; dk++) {
+                        int np = bp + dp;
+                        int nq = bq + dq;
+                        int nk = bk + dk;
+                        Chunk *c = find_chunk(np, nq, nk, 0);
+                        if(c && c->dirty) {
+                            gen_chunk_buffer(chunk);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+void place_block_global_cords(int x, int y, int z, int w) {
+    place_block(chunked_int(x), chunked_int(z), x, y, z, w);
+}
```

Comment:
Yes, I know that this function is a little weird. `block_place(...)` is just a copy-past of the inline code from `parse_buffer(...)` and I preferred to keep the changes at a minimum for now. I the future when the B protocol is removed we can re-factor and merge these functions.